daemon/containerd/image_delete.go

```diff
@@ -59,28 +63,269 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force,
+// isImageIDPrefix returns whether the given
+// possiblePrefix is a prefix of the given imageID.
+func isImageIDPrefix(imageID, possiblePrefix string) bool {
+	if strings.HasPrefix(imageID, possiblePrefix) {
+		return true
+	}
+	if i := strings.IndexRune(imageID, ':'); i >= 0 {
+		return strings.HasPrefix(imageID[i+1:], possiblePrefix)
+	}
+	return false
+}
+
+func sortParentsByAffinity(parents []imageWithRootfs) {
+	sort.Slice(parents, func(i, j int) bool {
+		return len(parents[i].rootfs.DiffIDs) > len(parents[j].rootfs.DiffIDs)
+	})
+}
+
+// isSingleReference returns true if there are no other images in the
+// daemon targeting the same content as `img` that are not dangling.
+func (i *ImageService) isSingleReference(ctx context.Context, img images.Image) (bool, error) {
+	refs, err := i.client.ImageService().List(ctx, "target.digest=="+img.Target.Digest.String())
+	if err != nil {
+		return false, err
+	}
+	for _, ref := range refs {
+		if !isDanglingImage(ref) && ref.Name != img.Name {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+type conflictType int
+
+const (
+	conflictRunningContainer conflictType = 1 << iota
+	conflictActiveReference
+	conflictStoppedContainer
+	conflictHard = conflictRunningContainer
+	conflictSoft = conflictActiveReference | conflictStoppedContainer
+)
+
+// imageDeleteHelper attempts to delete the given image from this daemon.
+// If the image has any hard delete conflicts (running containers using
+// the image) then it cannot be deleted. If the image has any soft delete
+// conflicts (any tags/digests referencing the image or any stopped container
+// using the image) then it can only be deleted if force is true. Any deleted
+// images and untagged references are appended to the given records. If any
+// error or conflict is encountered, it will be returned immediately without
+// deleting the image. If quiet is true, any encountered conflicts will be
+// ignored and the function will return nil immediately without deleting the image.
+func (i *ImageService) imageDeleteHelper(ctx context.Context, img images.Image, records *[]types.ImageDeleteResponseItem, force bool) error {
+	// First, determine if this image has any conflicts. Ignore soft conflicts
+	// if force is true.
+	c := conflictHard
+	if !force {
+		c |= conflictSoft
+	}
+
+	imgID := image.ID(img.Target.Digest)
+
+	conflict, err := i.checkImageDeleteConflict(ctx, imgID, c)
+	if err != nil {
+		return err
+	}
+	if conflict != nil {
+		// There was a conflict and it's either a hard conflict
+		// OR we are not forcing deletion on soft conflicts.
+		return conflict
+	}
+
+	refs, err := i.client.ImageService().List(ctx, "target.digest=="+imgID.String())
+	if err != nil {
+		return err
+	}
+	for _, eachRef := range refs {
+		err = i.client.ImageService().Delete(ctx, eachRef.Name, images.SynchronousDelete())
+		if err != nil {
+			return err
+		}
+		untaggedRef, err := reference.ParseAnyReference(eachRef.Name)
+		if err != nil {
+			return err
+		}
```

Comment:
Put the fallible operation which doesn't change the state of the system ahead of the operation which does.
```suggestion
		untaggedRef, err := reference.ParseAnyReference(eachRef.Name)
		if err != nil {
			return err
		}
		err = i.client.ImageService().Delete(ctx, eachRef.Name, images.SynchronousDelete())
		if err != nil {
			return err
		}
```